File: examples/workers/python/worker.py
#!/usr/bin/env python3
"""sched HTTP worker example — Python 3 stdlib (http.server), zero dependencies.

Implements the sched runner protocol (docs/content/docs/05.protocol.md):
  - POST /            dispatch envelope { task: { name, config }, data }
  - GET /status/:id   async poll endpoint (same outbound auth)

Tasks:
  - "ping"  sync:  200 { status: "succeeded"|"failed", ... }
  - "long"  async: 202 { status: "accepted", statusUrl, pollIntervalMs } ->
                    GET statusUrl -> running with progress -> terminal envelope

Auth (G1): when SCHED_API_KEY is set, every endpoint requires `x-sched-api-key`
to match (else 401). Unset -> validation skipped.
Idempotency: results are stored by `x-sched-run-id`; a redelivery of the same
run id answers with the saved outcome instead of re-executing.

Run: python3 worker.py   (Windows: py worker.py; port: SCHED_PORT, default 8081)
"""

import json
import os
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

API_KEY = os.environ.get("SCHED_API_KEY", "")
BASE_URL = os.environ.get("SCHED_BASE_URL", "").rstrip("/")
PORT = int(os.environ.get("SCHED_PORT", "8081"))

_lock = threading.Lock()
_runs = {}  # run_id -> state dict


def base_url(handler):
    if BASE_URL:
        return BASE_URL
    return "http://" + (handler.headers.get("Host") or "localhost")


def auth_ok(handler):
    return not API_KEY or handler.headers.get("x-sched-api-key") == API_KEY


def work_ms(data, default):
    try:
        ms = int(data.get("workMs", default))
        return ms if ms > 0 else default
    except (TypeError, ValueError):
        return default


def should_fail(data):
    return data.get("fail") is True


class Handler(BaseHTTPRequestHandler):
    def log_message(self, fmt, *args):  # quieter access log
        print("[worker] %s %s" % (self.command, self.path), flush=True)

    # -- plumbing ----------------------------------------------------------

    def _json(self, code, obj):
        body = json.dumps(obj).encode("utf-8")
        self.send_response(code)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _read_envelope(self):
        try:
            length = int(self.headers.get("content-length") or 0)
            raw = self.rfile.read(length) if length else b"{}"
            return json.loads(raw.decode("utf-8") or "{}")
        except (ValueError, json.JSONDecodeError):
            return None

    # -- endpoints ---------------------------------------------------------

    def do_POST(self):
        if not auth_ok(self):
            self._json(401, {"error": "unauthorized"})
            return
        env = self._read_envelope()
        if env is None:
            self._json(400, {"error": "invalid JSON envelope"})
            return
        run_id = self.headers.get("x-sched-run-id")
        if not run_id:
            run_id = "anon-%d" % time.time_ns()
        task_name = (env.get("task") or {}).get("name", "")
        if task_name == "long":
            self._async(env, run_id)
        else:
            self._sync(env, run_id)

    def do_GET(self):
        if not auth_ok(self):
            self._json(401, {"error": "unauthorized"})
            return
        if not self.path.startswith("/status/"):
            self._json(404, {"error": "not found"})
            return
        run_id = self.path[len("/status/"):]
        with _lock:
            state = _runs.get(run_id)
        if state is None:
            self._json(404, {"error": "unknown run id"})
            return
        self._json(200, state)

    # -- task handlers -----------------------------------------------------

    def _sync(self, env, run_id):
        data = env.get("data") or {}
        with _lock:
            existing = _runs.get(run_id)
        if existing is not None:  # redelivery -> saved outcome
            self._json(200, existing)
            return
        time.sleep(work_ms(data, 100) / 1000.0)
        if should_fail(data):
            state = {
                "status": "failed",
                "progress": 100,
                "error": "worker failed on purpose (data.fail=true)",
                "log": "ping failed as requested",
            }
        else:
            state = {
                "status": "succeeded",
                "progress": 100,
                "result": {"task": (env.get("task") or {}).get("name"), "data": data, "echo": "pong"},
                "log": "ping ok",
            }
        with _lock:
            _runs[run_id] = state
        self._json(200, state)

    def _async(self, env, run_id):
        data = env.get("data") or {}
        with _lock:
            existing = _runs.get(run_id)
            if existing is None:
                state = {"status": "queued", "progress": 0, "log": "accepted"}
                _runs[run_id] = state
        if existing is not None:  # redelivery of an in-flight or finished run
            if existing["status"] in ("queued", "running"):
                self._json(202, {"status": "accepted", "statusUrl": "%s/status/%s" % (base_url(self), run_id), "pollIntervalMs": 500})
            else:
                self._json(200, existing)
            return
        status_url = "%s/status/%s" % (base_url(self), run_id)
        ms = work_ms(data, 150)
        fail = should_fail(data)
        task_name = (env.get("task") or {}).get("name")

        def work():
            for p in range(20, 101, 20):
                time.sleep(ms / 1000.0)
                with _lock:
                    state["status"] = "running"
                    state["progress"] = p
                    state["log"] = "working… %d%%" % p
            with _lock:
                if fail:
                    state["status"] = "failed"
                    state["error"] = "worker failed on purpose (data.fail=true)"
                    state["log"] = "long task failed as requested"
                else:
                    state["status"] = "succeeded"
                    state["result"] = {"task": task_name, "workMs": ms, "ok": True}
                    state["log"] = "long task completed"
                state["progress"] = 100

        threading.Thread(target=work, daemon=True).start()
        self._json(202, {"status": "accepted", "statusUrl": status_url, "pollIntervalMs": 500})


def main():
    server = ThreadingHTTPServer(("0.0.0.0", PORT), Handler)
    print("[worker] sched example worker (Python) listening on :%d" % PORT, flush=True)
    server.serve_forever()


if __name__ == "__main__":
    main()

File: examples/workers/java/Worker.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;

/**
 * sched HTTP worker example — pure JDK (com.sun.net.httpserver), no build deps.
 *
 * Implements the sched runner protocol (docs/content/docs/05.protocol.md):
 *   - POST /            dispatch envelope { task: { name, config }, data }
 *   - GET /status/:id   async poll endpoint (same outbound auth)
 *
 * Tasks:
 *   - "ping"  sync:  200 { status: "succeeded"|"failed", ... }
 *   - "long"  async: 202 { status: "accepted", statusUrl, pollIntervalMs } ->
 *                    GET statusUrl -> running with progress -> terminal envelope
 *
 * Auth (G1): when SCHED_API_KEY is set, every endpoint requires `x-sched-api-key`
 * to match (else 401). Unset -> validation skipped.
 * Idempotency: results are stored by `x-sched-run-id`; a redelivery of the same
 * run id answers with the saved outcome instead of re-executing.
 *
 * Run: javac Worker.java && java Worker   (port: SCHED_PORT, default 8083)
 */
public class Worker {

    private static final Map<String, Map<String, Object>> RUNS = new ConcurrentHashMap<>();

    private static final String API_KEY = env("SCHED_API_KEY");
    private static final String BASE_URL = env("SCHED_BASE_URL").replaceAll("/+$", "");

    private static String env(String name) {
        String v = System.getenv(name);
        return v == null ? "" : v;
    }

    public static void main(String[] args) throws IOException {
        int port = 8083;
        if (!env("SCHED_PORT").isEmpty()) {
            try {
                port = Integer.parseInt(env("SCHED_PORT"));
            } catch (NumberFormatException ignored) {
            }
        }
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
        server.createContext("/", Worker::dispatch);
        server.createContext("/status/", Worker::status);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("[worker] sched example worker (Java) listening on :" + port);
    }

    // ---- endpoints ---------------------------------------------------------

    private static void dispatch(HttpExchange ex) throws IOException {
        if (!"POST".equals(ex.getRequestMethod())) {
            json(ex, 404, "{\"error\":\"not found\"}");
            return;
        }
        if (!authOk(ex)) {
            json(ex, 401, "{\"error\":\"unauthorized\"}");
            return;
        }
        Map<String, Object> env = readJson(ex);
        if (env == null) {
            json(ex, 400, "{\"error\":\"invalid JSON envelope\"}");
            return;
        }
        Map<String, Object> task = castMap(env.get("task"));
        String name = task == null ? "" : String.valueOf(task.get("name"));
        Map<String, Object> data = castMap(env.get("data"));
        if (data == null) {
            data = new HashMap<>();
        }
        String runId = ex.getRequestHeaders().getFirst("x-sched-run-id");
        if (runId == null || runId.isEmpty()) {
            runId = "anon-" + System.nanoTime();
        }
        if ("long".equals(name)) {
            async(ex, runId, name, data);
        } else {
            sync(ex, runId, name, data);
        }
    }

    private static void status(HttpExchange ex) throws IOException {
        if (!"GET".equals(ex.getRequestMethod())) {
            json(ex, 404, "{\"error\":\"not found\"}");
            return;
        }
        if (!authOk(ex)) {
            json(ex, 401, "{\"error\":\"unauthorized\"}");
            return;
        }
        String runId = ex.getRequestURI().getPath().substring("/status/".length());
        Map<String, Object> state = RUNS.get(runId);
        if (state == null) {
            json(ex, 404, "{\"error\":\"unknown run id\"}");
            return;
        }
        json(ex, 200, toJson(state));
    }

    // ---- task handlers -----------------------------------------------------

    private static void sync(HttpExchange ex, String runId, String taskName, Map<String, Object> data) throws IOException {
        Map<String, Object> existing = RUNS.get(runId); // redelivery -> saved outcome
        if (existing != null) {
            json(ex, 200, toJson(existing));
            return;
        }
        try {
            Thread.sleep(workMs(data, 100));
        } catch (InterruptedException ignored) {
        }
        Map<String, Object> state = new HashMap<>();
        state.put("progress", 100);
        if (shouldFail(data)) {
            state.put("status", "failed");
            state.put("error", "worker failed on purpose (data.fail=true)");
            state.put("log", "ping failed as requested");
        } else {
            Map<String, Object> result = new HashMap<>();
            result.put("task", taskName);
            result.put("data", data);
            result.put("echo", "pong");
            state.put("status", "succeeded");
            state.put("result", result);
            state.put("log", "ping ok");
        }
        RUNS.put(runId, state);
        json(ex, 200, toJson(state));
    }

    private static void async(HttpExchange ex, String runId, String taskName, Map<String, Object> data) throws IOException {
        Map<String, Object> existing = RUNS.get(runId); // redelivery of in-flight/finished run
        if (existing != null) {
            String st = String.valueOf(existing.get("status"));
            if ("queued".equals(st) || "running".equals(st)) {
                json(ex, 202, acceptedBody(ex, runId));
            } else {
                json(ex, 200, toJson(existing));
            }
            return;
        }
        final Map<String, Object> state = new HashMap<>();
        state.put("status", "queued");
        state.put("progress", 0);
        state.put("log", "accepted");
        RUNS.put(runId, state);
        final int ms = workMs(data, 150);
        final boolean fail = shouldFail(data);
        new Thread(() -> {
            try {
                for (int p = 20; p <= 100; p += 20) {
                    Thread.sleep(ms);
                    state.put("status", "running");
                    state.put("progress", p);
                    state.put("log", "working… " + p + "%");
                }
                state.put("progress", 100);
                if (fail) {
                    state.put("status", "failed");
                    state.put("error", "worker failed on purpose (data.fail=true)");
                    state.put("log", "long task failed as requested");
                } else {
                    Map<String, Object> result = new HashMap<>();
                    result.put("task", taskName);
                    result.put("workMs", ms);
                    result.put("ok", Boolean.TRUE);
                    state.put("status", "succeeded");
                    state.put("result", result);
                    state.put("log", "long task completed");
                }
            } catch (InterruptedException ignored) {
            }
        }).start();
        json(ex, 202, acceptedBody(ex, runId));
    }

    private static String acceptedBody(HttpExchange ex, String runId) {
        return "{\"status\":\"accepted\",\"statusUrl\":\"" + baseUrl(ex) + "/status/" + runId + "\",\"pollIntervalMs\":500}";
    }

    // ---- helpers -------------------------------------------------------------

    private static boolean authOk(HttpExchange ex) {
        return API_KEY.isEmpty() || API_KEY.equals(ex.getRequestHeaders().getFirst("x-sched-api-key"));
    }

    private static String baseUrl(HttpExchange ex) {
        if (!BASE_URL.isEmpty()) {
            return BASE_URL;
        }
        String host = ex.getRequestHeaders().getFirst("Host");
        if (host == null || host.isEmpty()) {
            host = "localhost:" + (env("SCHED_PORT").isEmpty() ? "8083" : env("SCHED_PORT"));
        }
        return "http://" + host;
    }

    private static int workMs(Map<String, Object> data, int def) {
        Object v = data.get("workMs");
        if (v instanceof Number) {
            int ms = ((Number) v).intValue();
            if (ms > 0) {
                return ms;
            }
        }
        return def;
    }

    private static boolean shouldFail(Map<String, Object> data) {
        return Boolean.TRUE.equals(data.get("fail"));
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> castMap(Object o) {
        return o instanceof Map ? (Map<String, Object>) o : null;
    }

    private static Map<String, Object> readJson(HttpExchange ex) throws IOException {
        ByteArrayOutputStream buf = new ByteArrayOutputStream();
        byte[] chunk = new byte[4096];
        int n;
        InputStream in = ex.getRequestBody();
        while ((n = in.read(chunk)) > 0) {
            buf.write(chunk, 0, n);
        }
        String text = new String(buf.toByteArray(), StandardCharsets.UTF_8);
        return parseJson(text);
    }

    private static void json(HttpExchange ex, int code, String body) throws IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("content-type", "application/json");
        ex.sendResponseHeaders(code, bytes.length);
        OutputStream os = ex.getResponseBody();
        os.write(bytes);
        os.close();
    }

    // ---- minimal JSON (JDK 8 has none in stdlib; tiny hand-rolled parser/serializer) ----

    private static String toJson(Object value) {
        StringBuilder sb = new StringBuilder();
        writeJson(sb, value);
        return sb.toString();
    }

    private static void writeJson(StringBuilder sb, Object v) {
        if (v == null) {
            sb.append("null");
        } else if (v instanceof String) {
            writeJsonString(sb, (String) v);
        } else if (v instanceof Boolean || v instanceof Number) {
            sb.append(v);
        } else if (v instanceof Map) {
            sb.append('{');
            boolean first = true;
            for (Map.Entry<?, ?> e : ((Map<?, ?>) v).entrySet()) {
                if (!first) {
                    sb.append(',');
                }
                first = false;
                writeJsonString(sb, String.valueOf(e.getKey()));
                sb.append(':');
                writeJson(sb, e.getValue());
            }
            sb.append('}');
        } else if (v instanceof Iterable) {
            sb.append('[');
            boolean first = true;
            for (Object o : (Iterable<?>) v) {
                if (!first) {
                    sb.append(',');
                }
                first = false;
                writeJson(sb, o);
            }
            sb.append(']');
        } else {
            writeJsonString(sb, String.valueOf(v));
        }
    }

    private static void writeJsonString(StringBuilder sb, String s) {
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':
                    sb.append("\\\"");
                    break;
                case '\\':
                    sb.append("\\\\");
                    break;
                case '\n':
                    sb.append("\\n");
                    break;
                case '\r':
                    sb.append("\\r");
                    break;
                case '\t':
                    sb.append("\\t");
                    break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append('"');
    }

    private static Map<String, Object> parseJson(String text) {
        Parser p = new Parser(text);
        Object v = p.parseValue();
        // strict: full consumption + no parse error + top-level object
        if (!p.ok || p.i < p.s.length()) {
            return null;
        }
        return v instanceof Map ? castMap(v) : null;
    }

    private static final class Parser {
        private final String s;
        private int i;
        private boolean ok = true;

        Parser(String s) {
            this.s = s;
        }

        Object parseValue() {
            skipWs();
            if (i >= s.length()) {
                ok = false; // empty / trailing input
                return null;
            }
            char c = s.charAt(i);
            if (c == '{') {
                return parseObject();
            }
            if (c == '[') {
                return parseArray();
            }
            if (c == '"') {
                return parseString();
            }
            if (c == 't') {
                expect("true");
                return Boolean.TRUE;
            }
            if (c == 'f') {
                expect("false");
                return Boolean.FALSE;
            }
            if (c == 'n') {
                expect("null");
                return null;
            }
            return parseNumber();
        }

        private void skipWs() {
            while (i < s.length() && Character.isWhitespace(s.charAt(i))) {
                i++;
            }
        }

        private Map<String, Object> parseObject() {
            Map<String, Object> m = new HashMap<>();
            i++; // {
            skipWs();
            if (i < s.length() && s.charAt(i) == '}') {
                i++;
                return m;
            }
            while (ok) {
                skipWs();
                if (i >= s.length() || s.charAt(i) != '"') {
                    ok = false;
                    return null;
                }
                String key = parseString();
                if (!ok) {
                    return null;
                }
                skipWs();
                if (i >= s.length() || s.charAt(i) != ':') {
                    ok = false;
                    return null;
                }
                i++;
                Object val = parseValue();
                if (!ok) {
                    return null;
                }
                m.put(key, val);
                skipWs();
                if (i >= s.length()) {
                    ok = false;
                    return null;
                }
                char c = s.charAt(i++);
                if (c == '}') {
                    return m;
                }
                if (c != ',') {
                    ok = false;
                    return null;
                }
            }
            return null;
        }

        private List<Object> parseArray() {
            List<Object> l = new ArrayList<>();
            i++; // [
            skipWs();
            if (i < s.length() && s.charAt(i) == ']') {
                i++;
                return l;
            }
            while (ok) {
                Object val = parseValue();
                if (!ok) {
                    return null;
                }
                l.add(val);
                skipWs();
                if (i >= s.length()) {
                    ok = false;
                    return null;
                }
                char c = s.charAt(i++);
                if (c == ']') {
                    return l;
                }
                if (c != ',') {
                    ok = false;
                    return null;
                }
            }
            return null;
        }

        private String parseString() {
            StringBuilder sb = new StringBuilder();
            i++; // opening quote
            while (i < s.length()) {
                char c = s.charAt(i++);
                if (c == '"') {
                    return sb.toString();
                }
                if (c == '\\' && i < s.length()) {
                    char e = s.charAt(i++);
                    switch (e) {
                        case '"':
                            sb.append('"');
                            break;
                        case '\\':
                            sb.append('\\');
                            break;
                        case '/':
                            sb.append('/');
                            break;
                        case 'n':
                            sb.append('\n');
                            break;
                        case 't':
                            sb.append('\t');
                            break;
                        case 'r':
                            sb.append('\r');
                            break;
                        case 'b':
                            sb.append('\b');
                            break;
                        case 'f':
                            sb.append('\f');
                            break;
                        case 'u':
                            if (i + 4 <= s.length()) {
                                sb.append((char) Integer.parseInt(s.substring(i, i + 4), 16));
                                i += 4;
                            } else {
                                ok = false; // truncated unicode escape
                            }
                            break;
                        default:
                            sb.append(e);
                    }
                } else if (c == '\\') {
                    ok = false; // trailing backslash
                } else {
                    sb.append(c);
                }
            }
            ok = false; // unterminated string
            return null;
        }

        private Number parseNumber() {
            int start = i;
            while (i < s.length() && "-+0123456789.eE".indexOf(s.charAt(i)) >= 0) {
                i++;
            }
            String num = s.substring(start, i);
            try {
                return Long.parseLong(num);
            } catch (NumberFormatException ignored) {
            }
            try {
                return Double.parseDouble(num);
            } catch (NumberFormatException e) {
                ok = false;
                return null;
            }
        }

        private void expect(String word) {
            if (s.startsWith(word, i)) {
                i += word.length();
            } else {
                ok = false;
            }
        }
    }
}
